go/worker/keymanager/churp.go

```diff
@@ -214,185 +232,129 @@ type submissionScheduler struct {
 	return &submissionScheduler{
-		logger:      logging.GetLogger("worker/keymanager/churp/submissions"),
+		logger:      logging.GetLogger("worker/keymanager/churp/submission"),
 		kmWorker:    kmWorker,
-		queue:       make([]*submissionInfo, 0),
-		running:     make(map[uint8]*submissionInfo),
-		submissions: make(map[uint8]*submissionInfo),
+		submissions: newTaskQueue(),
 	}
 }
 
 // Queue schedules the given CHURP scheme for application submission,
 // updating or removing the scheduled submission if already present.
 func (s *submissionScheduler) Queue(status *churp.Status) {
-	removeFn := func(info *submissionInfo, cause error) {
-		if info.cancel != nil {
-			info.cancel(cause)
-			delete(s.running, status.ID)
-		}
-		if info.index != -1 {
-			heap.Remove(&s.queue, info.index)
-		}
-		delete(s.submissions, status.ID)
-	}
-
 	// Stop and remove submission for the previous handoff.
-	info, ok := s.submissions[status.ID]
-	if ok && info.handoff < status.NextHandoff {
-		removeFn(info, fmt.Errorf("new handoff"))
+	info, ok := s.submissions.Get(status.ID)
+	if ok && info.status.NextHandoff < status.NextHandoff {
+		s.submissions.Remove(info, fmt.Errorf("new handoff"))
 	}
 
 	// Schedule submission for the current handoff.
-	info, ok = s.submissions[status.ID]
+	info, ok = s.submissions.Get(status.ID)
 	if !ok {
 		epoch := status.NextHandoff - 1
-		height, err := s.kmWorker.randomBlockHeight(epoch, 50)
+		height, err := s.kmWorker.selectBlockHeight(epoch, 20, 50)
 		if err != nil {
 			s.logger.Error("failed to select a random block height",
 				"err", err,
 			)
 			return
 		}
 
-		info = &submissionInfo{
-			churpID: status.ID,
-			handoff: status.NextHandoff,
-			height:  height,
-			index:   -1,
-			cancel:  nil,
-		}
-
-		s.submissions[status.ID] = info
-		heap.Push(&s.queue, info)
+		info = newTaskInfo(status, height)
+		s.submissions.Add(info)
 	}
 
 	// Stop and remove the submission if the application has already
 	// been submitted or if handoffs are disabled.
 	switch _, submitted := status.Applications[s.kmWorker.nodeID]; {
 	case submitted:
-		removeFn(info, fmt.Errorf("already submitted"))
+		s.submissions.Remove(info, fmt.Errorf("already submitted"))
 	case status.HandoffsDisabled():
-		removeFn(info, fmt.Errorf("handoffs disabled"))
+		s.submissions.Remove(info, fmt.Errorf("handoffs disabled"))
 	}
 }
 
 // Start starts all eligible queued submissions.
-func (s *submissionScheduler) Start(ctx context.Context, height int64) {
-	for {
-		if len(s.queue) == 0 {
-			return
-		}
-		info := s.queue.Peek().(*submissionInfo)
-		if info.height > height {
-			return
-		}
-		_ = heap.Pop(&s.queue)
-
-		submitCtx, submitCancel := context.WithCancelCause(ctx)
-
-		info.cancel = submitCancel
-		s.running[info.churpID] = info
+func (s *submissionScheduler) Start(height int64) {
+	for _, info := range s.submissions.Run(beacon.EpochMax, height) {
+		ctx, cancel := context.WithCancelCause(context.Background())
+		info.cancel = cancel
```

Comment:
What kind of synchronization did you have in mind?
Every node can participate in multiple instances and it needs to submit an application (a commitment) one epoch before the next handoff. These submissions don't need to be in sync, as instances are independent, but we do want that they happen at different (random) heights so that work is spread throughout the epoch.

Reply:
I meant thread/goroutine concurrency synchronization as I see these `info` structures possibly being accessed from different parts. Haven't yet looked into this too closely, it may be fine.

Reply:
`TaskInfo` structure is never accessed concurrently. All three go routines accept `Context` and `Status`, which fields don't change. So it should be safe.

Go routines:
```
s.submitApplication(ctx, info.status)
e.handoff(ctx, info.status, dimSwitchCh)
f.finalizeHandoff(ctx, status)
```